### PR TITLE
gh-103968: Deprecate creating heap types whose metaclass has custom tp_new.

### PR DESCRIPTION
**Pull request by encukou**
(That's a mouthful of an edge case!)

* Issue: gh-103968

### CODE REVIEW

File: Doc/whatsnew/3.12.rst [hunk 1]
```diff
@@ -1288,6 +1288,21 @@ Porting to Python 3.12
...
 
+* The following functions now select an appropriate metaclass for the newly
+  created type:
+
+  * :c:func:`PyType_FromSpec`
+  * :c:func:`PyType_FromSpecWithBases`
+  * :c:func:`PyType_FromModuleAndSpec`
+
+  Creating classes whose metaclass overrides :c:member:`~PyTypeObject.tp_new`
+  is deprecated, and in Python 3.14+ it will be no longer allowed.
```
Comment (barneygale) [inline Doc/whatsnew/3.12.rst]:
```suggestion
  is deprecated, and in Python 3.14+ it will be disallowed.
```

File: Doc/whatsnew/3.12.rst [hunk 2]
```diff
@@ -1364,6 +1379,11 @@ Deprecated
 * ``_PyErr_ChainExceptions`` is deprecated. Use ``_PyErr_ChainExceptions1``
   instead. (Contributed by Irit Katriel in :gh:`102192`.)
 
+* Using :c:func:`PyType_FromSpec`, :c:func:`PyType_FromSpecWithBases`
+  or :c:func:`PyType_FromModuleAndSpec` to create a class whose metaclass
+  overrides :c:member:`~PyTypeObject.tp_new`.
```
Comment (barneygale) [inline Doc/whatsnew/3.12.rst]:
```suggestion
  overrides :c:member:`~PyTypeObject.tp_new` is deprecated.
```

File: Doc/c-api/type.rst [hunk 3]
```diff
@@ -256,8 +256,13 @@ The following functions and structs are used to create
...
    (or *Py_tp_base[s]* slots if *bases* is ``NULL``, see below).
-   Note that metaclasses that override
-   :c:member:`~PyTypeObject.tp_new` are not supported.
+
+   Metaclasses that override :c:member:`~PyTypeObject.tp_new` are not
+   supported.
+   (For backwards compatibility, other ``PyType_From*`` functions allow
+   such metaclasses. They ignore ``tp_new``, which may result in incomplete
+   initialization. This is deprecated and in Python 3.14+ such metaclasses will
+   not be supported.)
```
Comment (barneygale) [inline Doc/c-api/type.rst:265]: This could be nested in a `.. deprecated:: 3.12` block I suppose? And below.

File: Objects/typeobject.c [hunk 4]
```diff
@@ -3634,8 +3634,9 @@ check_basicsize_includes_size_and_offsets(PyTypeObject* type)
 }
 
 PyObject *
-PyType_FromMetaclass(PyTypeObject *metaclass, PyObject *module,
-                     PyType_Spec *spec, PyObject *bases_in)
+_PyType_FromMetaclass_impl(
+    PyTypeObject *metaclass, PyObject *module,
+    PyType_Spec *spec, PyObject *bases_in, int _allow_tp_new)
```
Comment (erlend-aasland) [inline Objects/typeobject.c:3836]: That's an unusual naming; why the leading underscore in `_allow_tp_new`?

File: Objects/typeobject.c [hunk 5]
```diff
@@ -3634,8 +3634,9 @@ check_basicsize_includes_size_and_offsets(PyTypeObject* type)
 }
 
 PyObject *
-PyType_FromMetaclass(PyTypeObject *metaclass, PyObject *module,
-                     PyType_Spec *spec, PyObject *bases_in)
+_PyType_FromMetaclass_impl(
```
Comment (erlend-aasland) [inline Objects/typeobject.c:3834]: This function should be file local, so add the `static` keyword.

Also, file local helpers are normally lower-case only, and without a `Py` or `_Py` prefix.

File: Doc/c-api/type.rst (hunk 3 above)
Comment (encukou) [reply Doc/c-api/type.rst:265]: It's mentioned in `.. versionchanged:: 3.12` in the affected functions. 
Their docs are just ```` Equivalent to ``PyType_FromMetaclass(NULL, NULL, arg, NULL, NULL)`` ```` and a bunch of change notes, so the `PyType_FromMetaclass` docs need this parenthetical to be complete.

File: Objects/typeobject.c (hunk 5 above)
Comment (encukou) [reply Objects/typeobject.c:3834]: Thanks, I did forget `static`!

Trying out a few alternatives, I think this is the best way to say “implementation for `PyType_FromMetaclass`”, so I'd like to keep the name.
The file already has a few `static` functions with `_Py`.

File: Objects/typeobject.c (hunk 4 above)
Comment (encukou) [reply Objects/typeobject.c:3836]: It'll go away in a future version. It's even more unstable/internal than the surrounding code.

**Review by erlend-aasland — APPROVED** [review]

**Pull request merged**